dimod/include/dimod/presolver.h

```diff
@@ -0,0 +1,287 @@
+            original[p.first] = p.second;
+        }
+
+        // add the unfixed from variables_/reduced
+        for (size_type i = 0; i < variables_.size(); ++i) {
+            original[variables_[i]] = reduced[i];
+        }
+
+        return original;
+    }
+
+    void fix_variable(index_type v, assignment_type assignment) {
+        assert(v >= 0 && static_cast<size_type>(v) < variables_.size());
+        assignments_[variables_[v]] = assignment;
+        variables_.erase(variables_.begin() + v);
+    }
+
+    size_type num_variables_fixed() const { return this->assignments_.size(); }
+
+    size_type num_variables_unfixed() const { return this->variables_.size(); }
+
+ private:
+    std::vector<index_type> variables_;
+    std::unordered_map<index_type, assignment_type> assignments_;
+};
+
+template <class Bias, class Index = int, class Assignment = double>
+class PreSolver {
+ public:
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    /// Unsigned integral type that can represent non-negative values.
+    using size_type = std::size_t;
+
+    using assignment_type = Assignment;
+
+    using model_type = ConstrainedQuadraticModel<bias_type, index_type>;
+
+    using postsolver_type = PostSolver<bias_type, index_type, assignment_type>;
+
+ private:
+    model_type model_;
+
+    std::vector<std::unique_ptr<PreSolveBase<bias_type, index_type>>> presolvers_;
+
+    postsolver_type postsolver_;
+
+ public:
+    template <class B, class I>
+    explicit PreSolver(const ConstrainedQuadraticModel<B, I>& cqm)
+            : model_(cqm), postsolver_(cqm.num_variables()) {}
+
+    explicit PreSolver(model_type&& cqm)
+            : model_(std::move(cqm)), postsolver_(model_.num_variables()) {
+        assert(this->model_.num_variables() == this->postsolver_.num_variables_unfixed());
+    }
+
+    template <class T>
+    void add_presolver() {
+        this->presolvers_.emplace_back(
+                std::unique_ptr<PreSolveBase<bias_type, index_type>>(new T()));
+    }
+
+    void apply() {
+        for (auto&& presolver : this->presolvers_) {
+            presolver->apply(this);
+        }
+    }
+
+    void fix_variable(index_type v, assignment_type assignment) {
+        assert(v >= 0 && static_cast<size_type>(v) < this->model_.num_variables());
+        this->model_.fix_variable(v, assignment);
+        this->postsolver_.fix_variable(v, assignment);
+        assert(this->model_.num_variables() == this->postsolver_.num_variables_unfixed());
+    }
+
+    const model_type& model() const { return this->model_; }
+
+    const postsolver_type& postsolver() const { return this->postsolver_; }
+
+    /// Remove any variables where the upper and lower bounds are the same
+    size_type remove_fixed_variables() {
+        size_type start_size = this->model_.num_variables();
+        for (size_type v = 0; v < this->model_.num_variables();) {
```

Comment:
Why is v incremented in one case only ? May be a while loop will be more readable.